### ISSUE TITLE
LSP: Denols: cannot go to definition for cached imported dependencies: "index out of range" in _str_byteindex_enc

### ISSUE THREAD

**Issue by josephcrawfordSRH** [neovim/neovim#30719]
### Problem

When I go to a controller file for instance and try to `gd (vim.lsp.buf.definition)` to go to the source of the `BaseController` or any other imported dependency code, I get the following error.

```
Error executing vim.schedule lua callback: ...ar/neovim/0.10.1/share/nvim/runtime/lua/vim/lsp/util.lua:1914: index out of range
stack traceback:
    [C]: in function '_str_byteindex_enc'
    ...ar/neovim/0.10.1/share/nvim/runtime/lua/vim/lsp/util.lua:1914: in function 'locations_to_items'
    ...yVim/lazy/telescope.nvim/lua/telescope/builtin/__lsp.lua:205: in function 'handler'
    .../Cellar/neovim/0.10.1/share/nvim/runtime/lua/vim/lsp.lua:936: in function 'handler'
    .../neovim/0.10.1/share/nvim/runtime/lua/vim/lsp/client.lua:687: in function ''
    vim/_editor.lua: in function <vim/_editor.lua:0>
```

This is my configuration for the denols lsp

```lua
        denols = {
          cmd = { "/opt/homebrew/bin/deno", "lsp" },
          filetypes = { "typescript", "typescriptreact" },
          root_dir = function(...)
            return nvim_lsp.util.root_pattern("deno.jsonc", "deno.json")(...)
          end,
          settings = {
            deno = {
              enable = true,
              lint = true,
              unstable = true,
              suggest = {
                imports = {
                  hosts = {
                    ["https://deno.land"] = true,
                  },
                },
              },
            },
          },
        },
```

minimal_init.lua

```lua
--- CHANGE THESE
local pattern = '*.ts'
local cmd = "/opt/homebrew/bin/deno"
-- Add files/folders here that indicate the root of a project
local root_markers = {'deno.jsonc', 'deno.json'}
-- Change to table with settings if required
local settings = {
              enable = true,
              lint = true,
              unstable = true,
              suggest = {
                imports = {
                  hosts = {
                    ["https://deno.land"] = true,
                  },
                },
              }
            }

vim.api.nvim_create_autocmd('FileType', {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
    vim.lsp.start({
      name = 'bugged-ls',
      cmd = cmd,
      root_dir = root_dir,
      settings = settings
    })
  end
})
```

Any idea what I may be doing wrong here?

### Steps to reproduce using "nvim -u minimal_init.lua"

I have posted this to the deno discord server for assistance back on 10/02/2024 and no suggestions were given: https://discord.com/channels/684898665143206084/1291032358744752214/1291032358744752214

### Expected behavior

I would expect to be able to go to the definition of imported dependencies like you can in IDEs

### Nvim version (nvim -v)

0.1.0

### Language server name/version

denols 1.46.3

### Operating system/version

macos Sonoma 14.6.1

**Comment by tris203** [comment]
I think this would be helped by #30560 and #30585

**Comment by josephcrawfordSRH** [comment]
@tris203 so this is an actual issue that everyone is effected by and has to be lived with until those are released?  I just want to make sure I understand and stop looking for a solution if there is currently not one to be found :)

**Comment by tris203** [comment]
You can try setting ```offset_encoding = "utf-8"``` in the lsp setup which will bypass the ```str_byteindex()``` call. Note this is a work around rather than a fix

**Comment by josephcrawfordSRH** [comment]
@tris203 that seemed to get me closer, however, it seems to also take me to an empty file while throwing another error which is quite odd.

![Image](https://github.com/user-attachments/assets/4a57494b-5463-46b2-844c-75c6ef03679b)

**Comment by tris203** [comment]
This minimal init works for me:
```lua
local lsp = vim.lsp
local function buf_cache(bufnr, client)
  local params = {
    command = 'deno.cache',
    arguments = { {}, vim.uri_from_bufnr(bufnr) },
  }
  client.request('workspace/executeCommand', params, function(err, _result, ctx)
    if err then
      local uri = ctx.params.arguments[2]
      vim.api.nvim_err_writeln('cache command failed for ' .. vim.uri_to_fname(uri))
    end
  end, bufnr)
end

local function virtual_text_document_handler(uri, res, client)
  if not res then
    return nil
  end

  local lines = vim.split(res.result, '\n')
  local bufnr = vim.uri_to_bufnr(uri)

  local current_buf = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
  if #current_buf ~= 0 then
    return nil
  end

  vim.api.nvim_buf_set_lines(bufnr, 0, -1, false, lines)
  vim.api.nvim_set_option_value('readonly', true, { buf = bufnr })
  vim.api.nvim_set_option_value('modified', false, { buf = bufnr })
  vim.api.nvim_set_option_value('modifiable', false, { buf = bufnr })
  lsp.buf_attach_client(bufnr, client.id)
end

local function virtual_text_document(uri, client)
  local params = {
    textDocument = {
      uri = uri,
    },
  }
  local result = client.request_sync('deno/virtualTextDocument', params)
  virtual_text_document_handler(uri, result, client)
end

local function denols_handler(err, result, ctx, config)
  if not result or vim.tbl_isempty(result) then
    return nil
  end

  local client = vim.lsp.get_client_by_id(ctx.client_id)
  for _, res in pairs(result) do
    local uri = res.uri or res.targetUri
    if uri:match '^deno:' then
      virtual_text_document(uri, client)
      res['uri'] = uri
      res['targetUri'] = uri
    end
  end

  lsp.handlers[ctx.method](err, result, ctx, config)
end

--- CHANGE THESE
local pattern = 'typescript'
local cmd = { 'deno', 'lsp' }
-- Add files/folders here that indicate the root of a project
local root_markers = { 'deno.jsonc', 'deno.json' }
-- Change to table with settings if required
local settings = {
  enable = true,
  lint = true,
  unstable = true,
  suggest = {
    imports = {
      hosts = {
        ['https://deno.land'] = true,
      },
    },
  },
}

vim.api.nvim_create_autocmd('FileType', {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
    vim.lsp.start({
      name = 'bugged-ls',
      cmd = cmd,
      root_dir = root_dir,
      settings = settings,
      handlers = {
        ['textDocument/definition'] = denols_handler,
      },
    })
  end,
})
```

wrapper functions from here:
https://github.com/neovim/nvim-lspconfig/blob/04680101ff79e99b4e33a4386ec27cbd0d360c75/lua/lspconfig/configs/denols.lua#L47

**Comment by josephcrawfordSRH** [comment]
hmm interesting, I will have to cross reference the above with how LazyVim is implementing this as my configuration changes are minimal maybe it is an issue with how it is being loaded in LazyVim itself.

Thank you @tris203

**Comment by nicholascross** [comment]
> You can try setting `offset_encoding = "utf-8"` in the lsp setup which will bypass the `str_byteindex()` call. Note this is a work around rather than a fix

Thank you for this work around I was facing this issue when trying to use sourcekit-lsp and specifying the `offset_encoding` has resolved the issues I was facing with go to definition and other lsp operations.

**Comment by tris203** [comment]
I think this will have been fixed by #30747
